File: frontend/app/react-native/src/renderer/savedViews.ts
/**
 * Saved views for React Native listings — the RN counterpart of the web's `savedViewsStore`: a
 * named snapshot of the smart-search conditions (free-text `searchText` + every applied filter
 * value, range bounds included), with one optionally marked default (auto-applied on first load).
 * Parity note: the web persists in localStorage; this RN renderer has no AsyncStorage dependency, so
 * views are **session-scoped** (in-memory, keyed by the listing scope) and reset on app restart.
 */
export interface SavedView {
  name: string;
  /** filter values by state key (searchText + <fieldId> / <fieldId>_from / <fieldId>_to …) */
  values: Record<string, unknown>;
  isDefault?: boolean;
}

const store = new Map<string, SavedView[]>();

export const listSavedViews = (scope: string): SavedView[] => store.get(scope) ?? [];

/** Save (or replace, matching by name) a view. Empty names / empty snapshots are ignored. */
export const saveView = (scope: string, view: SavedView): void => {
  const name = view.name?.trim();
  if (!name || Object.keys(view.values ?? {}).length === 0) return;
  const views = listSavedViews(scope).filter((v) => v.name !== name);
  views.push({ ...view, name });
  store.set(scope, views);
};

export const deleteView = (scope: string, name: string): void => {
  const views = listSavedViews(scope).filter((v) => v.name !== name);
  if (views.length === 0) store.delete(scope);
  else store.set(scope, views);
};

/** Toggle the default flag on `name` (clearing it from every other view in the scope). */
export const setDefaultView = (scope: string, name: string): void => {
  const views = listSavedViews(scope).map((v) => ({
    ...v,
    isDefault: v.name === name ? !v.isDefault : false,
  }));
  store.set(scope, views);
};

export const defaultView = (scope: string): SavedView | undefined =>
  listSavedViews(scope).find((v) => v.isDefault);

File: frontend/app/intellij-plugin/src/main/kotlin/io/mateu/ijp/ui/SavedViews.kt
package io.mateu.ijp.ui

import com.fasterxml.jackson.databind.ObjectMapper
import com.intellij.ide.util.PropertiesComponent

/**
 * Saved views for a crud listing — parity with the web `savedViewsStore`: named snapshots of the
 * search conditions (searchText + filter state keys, range bounds included), one optionally the
 * default (auto-applied on first load), persisted per crud route in [PropertiesComponent].
 *
 * Serialized as a JSON array of plain maps via a Java-shaped ObjectMapper (no Kotlin databind module
 * needed): each entry is {name, values(map), isDefault}.
 */
object SavedViews {

    private val mapper = ObjectMapper()
    private fun key(scope: String) = "mateu.views.${scope.ifBlank { "_" }}"

    class View(val name: String, val values: Map<String, Any?>, val isDefault: Boolean)

    fun list(scope: String): List<View> {
        val raw = PropertiesComponent.getInstance().getValue(key(scope)) ?: return emptyList()
        return runCatching {
            (mapper.readValue(raw, List::class.java) as List<*>).mapNotNull { e ->
                val m = e as? Map<*, *> ?: return@mapNotNull null
                val name = m["name"] as? String ?: return@mapNotNull null
                @Suppress("UNCHECKED_CAST")
                View(name, (m["values"] as? Map<String, Any?>) ?: emptyMap(), m["isDefault"] == true)
            }
        }.getOrDefault(emptyList())
    }

    private fun persist(scope: String, views: List<View>) {
        val pc = PropertiesComponent.getInstance()
        if (views.isEmpty()) {
            pc.unsetValue(key(scope))
            return
        }
        val arr = views.map { mapOf("name" to it.name, "values" to it.values, "isDefault" to it.isDefault) }
        pc.setValue(key(scope), mapper.writeValueAsString(arr))
    }

    /** Save (or replace, matching by name) a view. Empty name / empty snapshot ignored. */
    fun save(scope: String, name: String, values: Map<String, Any?>) {
        val n = name.trim()
        if (n.isEmpty() || values.isEmpty()) return
        val views = list(scope).filter { it.name != n }.toMutableList()
        views.add(View(n, values, false))
        persist(scope, views)
    }

    fun delete(scope: String, name: String) = persist(scope, list(scope).filter { it.name != name })

    /** Toggle the default flag on [name], clearing it from every other view in the scope. */
    fun toggleDefault(scope: String, name: String) =
        persist(scope, list(scope).map { View(it.name, it.values, if (it.name == name) !it.isDefault else false) })

    fun default(scope: String): View? = list(scope).firstOrNull { it.isDefault }
}
